File: mobile/app/src/main/java/com/odyssey/travelplanner/data/WeatherRepository.kt
package com.odyssey.travelplanner.data

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.request.get
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.supervisorScope
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.net.URLEncoder
import java.util.Locale

data class WeatherSnapshot(
    val temperature: String,
    val condition: String,
    val tripTemperature: String? = null,
    val tripCondition: String? = null,
    val tripIsEstimate: Boolean = false,
    val tripDays: Map<String, WeatherDaySnapshot> = emptyMap(),
)

data class WeatherDaySnapshot(
    val temperature: String? = null,
    val condition: String? = null,
    val isEstimate: Boolean = false,
)

@Serializable
private data class OpenMeteoResponse(
    val current: OpenMeteoCurrent? = null,
    val daily: OpenMeteoDaily? = null,
)

@Serializable
private data class OpenMeteoArchiveResponse(
    val daily: OpenMeteoDaily? = null,
)

@Serializable
private data class OpenMeteoClimateResponse(
    val daily: OpenMeteoDaily? = null,
)

@Serializable
private data class OpenMeteoGeocodingResponse(
    val results: List<OpenMeteoGeocodingResult>? = null,
)

@Serializable
private data class OpenMeteoGeocodingResult(
    val latitude: Double,
    val longitude: Double,
)

@Serializable
private data class OpenMeteoCurrent(
    val temperature_2m: Double? = null,
    val weather_code: Int? = null,
)

@Serializable
private data class OpenMeteoDaily(
    val time: List<String> = emptyList(),
    val temperature_2m_mean: List<Double?> = emptyList(),
    val temperature_2m_max: List<Double?> = emptyList(),
    val weather_code: List<Int?> = emptyList(),
    val precipitation_sum: List<Double?> = emptyList(),
    val cloud_cover_mean: List<Double?> = emptyList(),
)

private data class TripDayWeather(
    val temperature: String?,
    val condition: String?,
    val isEstimate: Boolean = false,
)

private const val OPTIONAL_TRIP_WEATHER_TIMEOUT_MILLIS = 4_000L

class WeatherRepository {
    private val http = HttpClient(OkHttp) {
        expectSuccess = true
        install(HttpTimeout) {
            connectTimeoutMillis = 10_000
            requestTimeoutMillis = 15_000
            socketTimeoutMillis = 15_000
        }
        install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
    }

    fun close() {
        http.close()
    }

    suspend fun loadCurrent(
        cities: List<String>,
        tripDates: String = "",
        cityCoordinates: Map<String, CityLocation> = emptyMap(),
    ): Map<String, WeatherSnapshot> {
        val tripDateRange = tripDateRangeFrom(tripDates)
        val targetDate = tripDateRange?.first
        val tripDatesToLoad = tripDateRange?.let(::datesBetween) ?: emptyList()
        return supervisorScope {
            cities
                .filter(String::isNotBlank)
                .distinctBy { it.trim().lowercase(Locale.ROOT) }
                .map { city ->
                async {
                    val coordinates = cityCoordinates[city]?.let { it.latitude to it.longitude }
                        ?: cityCatalogEntry(city)?.let { it.latitude to it.longitude }
                        ?: resolveCoordinates(city)
                        ?: return@async null
                    val weather = requestOrNull<OpenMeteoResponse> {
                        http.get(
                            "https://api.open-meteo.com/v1/forecast?latitude=${coordinates.first}&longitude=${coordinates.second}&current=temperature_2m,weather_code&daily=temperature_2m_max,weather_code&forecast_days=16&timezone=auto",
                        ).body<OpenMeteoResponse>()
                    } ?: return@async null
                    val current = weather.current ?: return@async null
                    val currentTemperature = current.temperature_2m ?: return@async null
                    val currentWeatherCode = current.weather_code ?: return@async null
                    val tripDays = weather.daily.toTripDays().toMutableMap()
                    val missingTripDates = tripDatesToLoad.filterNot { date ->
                        tripDays.containsKey(date.toString())
                    }
                    val today = LocalDate.now()
                    val archivedDates = missingTripDates.filter { it.isBefore(today) }
                    if (archivedDates.isNotEmpty()) {
                        withTimeoutOrNull(OPTIONAL_TRIP_WEATHER_TIMEOUT_MILLIS) {
                            loadArchivedTripWeather(coordinates, archivedDates.first(), archivedDates.last())
                        }.orEmpty().forEach { (date, tripWeather) ->
                            tripDays[date] = WeatherDaySnapshot(
                                temperature = tripWeather.temperature,
                                condition = tripWeather.condition,
                                isEstimate = tripWeather.isEstimate,
                            )
                        }
                    }
                    val climateDates = missingTripDates.filter { it.isAfter(today.plusDays(15)) }
                    if (climateDates.isNotEmpty()) {
                        withTimeoutOrNull(OPTIONAL_TRIP_WEATHER_TIMEOUT_MILLIS) {
                            loadClimateTripWeather(coordinates, climateDates.first(), climateDates.last())
                        }.orEmpty().forEach { (date, tripWeather) ->
                            tripDays[date] = WeatherDaySnapshot(
                                temperature = tripWeather.temperature,
                                condition = tripWeather.condition,
                                isEstimate = tripWeather.isEstimate,
                            )
                        }
                    }
                    val firstTripDay = targetDate?.let { tripDays[it.toString()] }
                    city to WeatherSnapshot(
                        temperature = "${currentTemperature.toInt()}°C",
                        condition = conditionFor(currentWeatherCode),
                        tripTemperature = firstTripDay?.temperature,
                        tripCondition = firstTripDay?.condition,
                        tripIsEstimate = firstTripDay?.isEstimate == true,
                        tripDays = tripDays,
                    )
                }
            }.awaitAll().filterNotNull().toMap()
        }
    }

    private suspend fun resolveCoordinates(city: String): Pair<Double, Double>? {
        val query = city.trim()
            .replace(" — ", ", ")
            .takeIf { it.length >= 2 }
            ?: return null
        val encodedQuery = URLEncoder.encode(query, "UTF-8")
        val response = requestOrNull<OpenMeteoGeocodingResponse> {
            http.get(
                "https://geocoding-api.open-meteo.com/v1/search?name=$encodedQuery&count=1&language=en&format=json",
            ).body<OpenMeteoGeocodingResponse>()
        } ?: return null
        return response.results?.firstOrNull()?.let { it.latitude to it.longitude }
    }

    private suspend fun loadArchivedTripWeather(
        coordinates: Pair<Double, Double>,
        startDate: LocalDate,
        endDate: LocalDate,
    ): Map<String, TripDayWeather> {
        val archive = requestOrNull<OpenMeteoArchiveResponse> {
            http.get(
                "https://archive-api.open-meteo.com/v1/archive?latitude=${coordinates.first}&longitude=${coordinates.second}&start_date=$startDate&end_date=$endDate&daily=temperature_2m_max,weather_code&timezone=auto",
            ).body<OpenMeteoArchiveResponse>()
        } ?: return emptyMap()
        return archive.daily.toTripWeatherDays()
    }

    private suspend fun loadClimateTripWeather(
        coordinates: Pair<Double, Double>,
        startDate: LocalDate,
        endDate: LocalDate,
    ): Map<String, TripDayWeather> {
        val climate = requestOrNull<OpenMeteoClimateResponse> {
            http.get(
                "https://climate-api.open-meteo.com/v1/climate?latitude=${coordinates.first}&longitude=${coordinates.second}&start_date=$startDate&end_date=$endDate&models=EC_Earth3P_HR&daily=temperature_2m_mean,precipitation_sum,cloud_cover_mean&timezone=auto",
            ).body<OpenMeteoClimateResponse>()
        } ?: return emptyMap()
        return climate.daily.toClimateTripWeatherDays()
    }

    private suspend fun <T> requestOrNull(request: suspend () -> T): T? = try {
        request()
    } catch (cancellation: CancellationException) {
        throw cancellation
    } catch (_: Throwable) {
        null
    }
}

internal fun tripDateRangeFrom(value: String): Pair<LocalDate, LocalDate>? {
    val dotted = Regex("\\d{1,2}\\.\\d{1,2}\\.\\d{4}").find(value)?.value
    if (dotted != null) {
        val dates = Regex("\\d{1,2}\\.\\d{1,2}\\.\\d{4}").findAll(value).mapNotNull { match ->
            val parts = match.value.split('.')
            runCatching { LocalDate.of(parts[2].toInt(), parts[1].toInt(), parts[0].toInt()) }.getOrNull()
        }.toList()
        val start = dates.firstOrNull() ?: return null
        return start to dates.getOrElse(1) { start }
    }

    val isoDates = Regex("\\d{4}-\\d{2}-\\d{2}").findAll(value).mapNotNull { match ->
        runCatching { LocalDate.parse(match.value) }.getOrNull()
    }.toList()
    if (isoDates.isNotEmpty()) {
        val start = isoDates.first()
        return start to isoDates.getOrElse(1) { start }
    }

    val russianMonths = mapOf(
        "января" to 1, "январь" to 1,
        "февраля" to 2, "февраль" to 2,
        "марта" to 3, "март" to 3,
        "апреля" to 4, "апрель" to 4,
        "мая" to 5, "май" to 5,
        "июня" to 6, "июнь" to 6,
        "июля" to 7, "июль" to 7,
        "августа" to 8, "август" to 8,
        "сентября" to 9, "сентябрь" to 9,
        "октября" to 10, "октябрь" to 10,
        "ноября" to 11, "ноябрь" to 11,
        "декабря" to 12, "декабрь" to 12,
    )
    val compactHumanRange = Regex(
        "(\\d{1,2})\\s*[–—-]\\s*(\\d{1,2})\\s+(${russianMonths.keys.joinToString("|")})\\s+(\\d{4})",
        RegexOption.IGNORE_CASE,
    ).find(value)
    if (compactHumanRange != null) {
        val startDay = compactHumanRange.groupValues[1].toIntOrNull() ?: return null
        val endDay = compactHumanRange.groupValues[2].toIntOrNull() ?: return null
        val month = russianMonths[compactHumanRange.groupValues[3].lowercase(Locale.ROOT)] ?: return null
        val year = compactHumanRange.groupValues[4].toIntOrNull() ?: return null
        val start = runCatching { LocalDate.of(year, month, startDay) }.getOrNull() ?: return null
        val end = runCatching { LocalDate.of(year, month, endDay) }.getOrNull() ?: return null
        return start to end
    }
    val dates = Regex(
        "(\\d{1,2})\\s+(${russianMonths.keys.joinToString("|")})\\s+(\\d{4})",
        RegexOption.IGNORE_CASE,
    ).findAll(value).mapNotNull { match ->
        val day = match.groupValues[1].toIntOrNull() ?: return@mapNotNull null
        val month = russianMonths[match.groupValues[2].lowercase(Locale.ROOT)] ?: return@mapNotNull null
        val year = match.groupValues[3].toIntOrNull() ?: return@mapNotNull null
        runCatching { LocalDate.of(year, month, day) }.getOrNull()
    }.toList()
    val start = dates.firstOrNull() ?: return null
    return start to dates.getOrElse(1) { start }
}

private fun datesBetween(range: Pair<LocalDate, LocalDate>, maxDays: Int = 366): List<LocalDate> {
    val dayCount = ChronoUnit.DAYS.between(range.first, range.second)
    if (dayCount < 0) return emptyList()
    val lastOffset = dayCount.coerceAtMost((maxDays - 1).coerceAtLeast(0).toLong())
    return (0..lastOffset).map { offset -> range.first.plusDays(offset) }
}

private fun OpenMeteoDaily?.toTripDays(): Map<String, WeatherDaySnapshot> {
    val daily = this ?: return emptyMap()
    return daily.time.mapIndexedNotNull { index, date ->
        val temperature = daily.temperature_2m_max.getOrNull(index)?.let { "${it.toInt()}°C" }
        val condition = daily.weather_code.getOrNull(index)?.let(::conditionFor)
        if (temperature == null && condition == null) {
            null
        } else {
            date to WeatherDaySnapshot(temperature = temperature, condition = condition)
        }
    }.toMap()
}

private fun OpenMeteoDaily?.toTripWeatherDays(): Map<String, TripDayWeather> {
    val daily = this ?: return emptyMap()
    return daily.time.mapIndexedNotNull { index, date ->
        val temperature = daily.temperature_2m_max.getOrNull(index)?.let { "${it.toInt()}°C" }
        val condition = daily.weather_code.getOrNull(index)?.let(::conditionFor)
        if (temperature == null && condition == null) {
            null
        } else {
            date to TripDayWeather(temperature, condition)
        }
    }.toMap()
}

private fun OpenMeteoDaily?.toClimateTripWeatherDays(): Map<String, TripDayWeather> {
    val daily = this ?: return emptyMap()
    return daily.time.mapIndexedNotNull { index, date ->
        val temperature = daily.temperature_2m_mean.getOrNull(index)?.let { "${it.toInt()}°C" }
        val precipitation = daily.precipitation_sum.getOrNull(index) ?: 0.0
        val cloudCover = daily.cloud_cover_mean.getOrNull(index) ?: 0.0
        val condition = when {
            precipitation >= 1.0 -> "Дождь"
            cloudCover >= 65.0 -> "Облачно"
            else -> "Ясно"
        }
        if (temperature == null && condition.isBlank()) {
            null
        } else {
            date to TripDayWeather(temperature, condition, isEstimate = true)
        }
    }.toMap()
}

private fun conditionFor(code: Int): String = when (code) {
    0 -> "Ясно"
    1, 2, 3 -> "Облачно"
    45, 48 -> "Туман"
    51, 53, 55, 56, 57 -> "Морось"
    61, 63, 65, 66, 67, 80, 81, 82 -> "Дождь"
    71, 73, 75, 77, 85, 86 -> "Снег"
    95, 96, 99 -> "Гроза"
    else -> "—"
}

File: ios/Ramingo/Networking/WeatherRepository.swift
import Foundation

struct WeatherSnapshot: Hashable, Sendable {
    let temperature: String
    let condition: String
    let tripTemperature: String?
    let tripCondition: String?
    let tripIsEstimate: Bool
    let tripDays: [String: WeatherDaySnapshot]
}

struct WeatherDaySnapshot: Hashable, Sendable {
    let temperature: String?
    let condition: String?
    let isEstimate: Bool
}

private struct TripWeather {
    let temperature: String
    let condition: String
    let isEstimate: Bool
}

private struct WeatherResponse: Decodable {
    let current: CurrentWeather
    let daily: DailyWeather?
}

private struct CurrentWeather: Decodable {
    let temperature: Double
    let weatherCode: Int

    enum CodingKeys: String, CodingKey {
        case temperature = "temperature_2m"
        case weatherCode = "weather_code"
    }
}

private struct DailyWeather: Decodable {
    let time: [String]
    let maxTemperature: [Double]
    let meanTemperature: [Double]
    let weatherCodes: [Int]
    let precipitation: [Double]
    let cloudCover: [Double]

    enum CodingKeys: String, CodingKey {
        case time
        case maxTemperature = "temperature_2m_max"
        case meanTemperature = "temperature_2m_mean"
        case weatherCodes = "weather_code"
        case precipitation = "precipitation_sum"
        case cloudCover = "cloud_cover_mean"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        time = try container.decodeIfPresent([String].self, forKey: .time) ?? []
        maxTemperature = try container.decodeIfPresent([Double].self, forKey: .maxTemperature) ?? []
        meanTemperature = try container.decodeIfPresent([Double].self, forKey: .meanTemperature) ?? []
        weatherCodes = try container.decodeIfPresent([Int].self, forKey: .weatherCodes) ?? []
        precipitation = try container.decodeIfPresent([Double].self, forKey: .precipitation) ?? []
        cloudCover = try container.decodeIfPresent([Double].self, forKey: .cloudCover) ?? []
    }
}

private struct DailyOnlyResponse: Decodable {
    let daily: DailyWeather?
}

private struct GeocodingResponse: Decodable {
    let results: [GeocodingResult]?
}

private struct GeocodingResult: Decodable {
    let latitude: Double
    let longitude: Double
}

final class WeatherRepository {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func loadCurrent(
        cities: [String],
        tripDates: String,
        coordinates: [String: Coordinate],
    ) async -> [String: WeatherSnapshot] {
        let uniqueCities = cities.reduce(into: [String]()) { result, city in
            let value = city.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !value.isEmpty,
                  !result.contains(where: { $0.caseInsensitiveCompare(value) == .orderedSame })
            else { return }
            result.append(value)
        }

        return await withTaskGroup(of: (String, WeatherSnapshot)?.self) { group in
            for city in uniqueCities {
                group.addTask {
                    guard let coordinate = await self.resolveCoordinate(for: city, coordinates: coordinates),
                          let snapshot = try? await self.load(city: city, coordinate: coordinate, tripDates: tripDates)
                    else { return nil }
                    return (city, snapshot)
                }
            }

            var result: [String: WeatherSnapshot] = [:]
            for await item in group {
                if let item {
                    result[item.0] = item.1
                }
            }
            return result
        }
    }

    func resolveCoordinates(
        cities: [String],
        known: [String: Coordinate],
    ) async -> [String: Coordinate] {
        var result = known
        for city in cities.uniqued() where !city.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if result[city] != nil { continue }
            if let coordinate = await resolveCoordinate(for: city, coordinates: result) {
                result[city] = coordinate
            }
        }
        return result
    }

    private func load(city: String, coordinate: Coordinate, tripDates: String) async throws -> WeatherSnapshot {
        var components = URLComponents(string: "https://api.open-meteo.com/v1/forecast")!
        components.queryItems = [
            URLQueryItem(name: "latitude", value: String(coordinate.latitude)),
            URLQueryItem(name: "longitude", value: String(coordinate.longitude)),
            URLQueryItem(name: "current", value: "temperature_2m,weather_code"),
            URLQueryItem(name: "daily", value: "temperature_2m_max,weather_code"),
            URLQueryItem(name: "forecast_days", value: "16"),
            URLQueryItem(name: "timezone", value: "auto"),
        ]
        let (data, response) = try await requestData(from: components.url!)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw WeatherError.requestFailed }
        let weather = try JSONDecoder().decode(WeatherResponse.self, from: data)
        let tripDateRange = iosWeatherTripDateRange(tripDates)
        let targetDate = tripDateRange?.0
        let tripDatesToLoad = tripDateRange.map { iosWeatherTripDates(tripDates) } ?? []
        var tripDays = weatherTripDays(from: weather.daily)
        let missingTripDates = tripDatesToLoad.filter { tripDays[iosWeatherISODate($0)] == nil }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .current
        let today = calendar.startOfDay(for: Date())
        let forecastEnd = calendar.date(byAdding: .day, value: 15, to: today) ?? today

        let archivedDates = missingTripDates.filter { $0 < today }
        if !archivedDates.isEmpty,
           let archived = try? await loadArchivedTripWeather(
               coordinate: coordinate,
               startDate: archivedDates[0],
               endDate: archivedDates[archivedDates.count - 1],
           ) {
            for (date, trip) in archived {
                tripDays[date] = WeatherDaySnapshot(
                    temperature: trip.temperature,
                    condition: trip.condition,
                    isEstimate: trip.isEstimate,
                )
            }
        }

        let climateDates = missingTripDates.filter { $0 > forecastEnd }
        if !climateDates.isEmpty,
           let climate = try? await loadClimateTripWeather(
               coordinate: coordinate,
               startDate: climateDates[0],
               endDate: climateDates[climateDates.count - 1],
           ) {
            for (date, trip) in climate {
                tripDays[date] = WeatherDaySnapshot(
                    temperature: trip.temperature,
                    condition: trip.condition,
                    isEstimate: trip.isEstimate,
                )
            }
        }

        let firstTripDay = targetDate.flatMap { tripDays[iosWeatherISODate($0)] }
        return WeatherSnapshot(
            temperature: "\(Int(weather.current.temperature.rounded()))°C",
            condition: condition(for: weather.current.weatherCode),
            tripTemperature: firstTripDay?.temperature,
            tripCondition: firstTripDay?.condition,
            tripIsEstimate: firstTripDay?.isEstimate == true,
            tripDays: tripDays,
        )
    }

    private func resolveCoordinate(for city: String, coordinates: [String: Coordinate]) async -> Coordinate? {
        if let exact = coordinates[city] { return exact }
        if let matching = coordinates.first(where: { $0.key.caseInsensitiveCompare(city) == .orderedSame })?.value {
            return matching
        }
        var components = URLComponents(string: "https://geocoding-api.open-meteo.com/v1/search")!
        components.queryItems = [
            URLQueryItem(name: "name", value: city.replacingOccurrences(of: " — ", with: ", ")),
            URLQueryItem(name: "count", value: "1"),
            URLQueryItem(name: "language", value: "en"),
            URLQueryItem(name: "format", value: "json"),
        ]
        guard let url = components.url,
              let (data, response) = try? await requestData(from: url),
              (response as? HTTPURLResponse)?.statusCode == 200,
              let decoded = try? JSONDecoder().decode(GeocodingResponse.self, from: data),
              let result = decoded.results?.first
        else { return nil }
        return Coordinate(latitude: result.latitude, longitude: result.longitude)
    }

    private func loadArchivedTripWeather(
        coordinate: Coordinate,
        startDate: Date,
        endDate: Date,
    ) async throws -> [String: TripWeather] {
        let day = ISO8601DateFormatter()
        day.formatOptions = [.withFullDate]
        var components = URLComponents(string: "https://archive-api.open-meteo.com/v1/archive")!
        components.queryItems = [
            URLQueryItem(name: "latitude", value: String(coordinate.latitude)),
            URLQueryItem(name: "longitude", value: String(coordinate.longitude)),
            URLQueryItem(name: "start_date", value: day.string(from: startDate)),
            URLQueryItem(name: "end_date", value: day.string(from: endDate)),
            URLQueryItem(name: "daily", value: "temperature_2m_max,weather_code"),
            URLQueryItem(name: "timezone", value: "auto"),
        ]
        let (data, response) = try await requestData(from: components.url!)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw WeatherError.requestFailed }
        let archive = try JSONDecoder().decode(DailyOnlyResponse.self, from: data)
        return tripWeatherDays(in: archive.daily)
    }

    private func loadClimateTripWeather(
        coordinate: Coordinate,
        startDate: Date,
        endDate: Date,
    ) async throws -> [String: TripWeather] {
        let day = ISO8601DateFormatter()
        day.formatOptions = [.withFullDate]
        var components = URLComponents(string: "https://climate-api.open-meteo.com/v1/climate")!
        components.queryItems = [
            URLQueryItem(name: "latitude", value: String(coordinate.latitude)),
            URLQueryItem(name: "longitude", value: String(coordinate.longitude)),
            URLQueryItem(name: "start_date", value: day.string(from: startDate)),
            URLQueryItem(name: "end_date", value: day.string(from: endDate)),
            URLQueryItem(name: "models", value: "EC_Earth3P_HR"),
            URLQueryItem(name: "daily", value: "temperature_2m_mean,precipitation_sum,cloud_cover_mean"),
            URLQueryItem(name: "timezone", value: "auto"),
        ]
        let (data, response) = try await requestData(from: components.url!)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw WeatherError.requestFailed }
        let climate = try JSONDecoder().decode(DailyOnlyResponse.self, from: data)
        guard let daily = climate.daily else { throw WeatherError.requestFailed }
        var result: [String: TripWeather] = [:]
        for (index, date) in daily.time.enumerated() {
            let temperature = daily.meanTemperature[safe: index].map { "\(Int($0.rounded()))°C" }
            let precipitation = daily.precipitation[safe: index] ?? 0
            let cloudCover = daily.cloudCover[safe: index] ?? 0
            let condition = precipitation >= 1 ? "Дождь" : cloudCover >= 65 ? "Облачно" : "Ясно"
            if temperature != nil || !condition.isEmpty {
                result[date] = TripWeather(
                    temperature: temperature ?? "—",
                    condition: condition,
                    isEstimate: true,
                )
            }
        }
        return result
    }

    private func tripWeatherDays(in daily: DailyWeather?) -> [String: TripWeather] {
        guard let daily else { return [:] }
        var result: [String: TripWeather] = [:]
        for (index, date) in daily.time.enumerated() {
            let temperature = daily.maxTemperature[safe: index].map { "\(Int($0.rounded()))°C" }
            let condition = daily.weatherCodes[safe: index].map { self.condition(for: $0) }
            if temperature != nil || condition != nil {
                result[date] = TripWeather(
                    temperature: temperature ?? "—",
                    condition: condition ?? "—",
                    isEstimate: false,
                )
            }
        }
        return result
    }

    private func weatherTripDays(from daily: DailyWeather?) -> [String: WeatherDaySnapshot] {
        guard let daily else { return [:] }
        var result: [String: WeatherDaySnapshot] = [:]
        for (index, date) in daily.time.enumerated() {
            let temperature = daily.maxTemperature[safe: index].map { "\(Int($0.rounded()))°C" }
            let condition = daily.weatherCodes[safe: index].map { self.condition(for: $0) }
            if temperature != nil || condition != nil {
                result[date] = WeatherDaySnapshot(temperature: temperature, condition: condition, isEstimate: false)
            }
        }
        return result
    }

    private func condition(for code: Int) -> String {
        switch code {
        case 0: return "Ясно"
        case 1, 2, 3: return "Облачно"
        case 45, 48: return "Туман"
        case 51, 53, 55, 56, 57: return "Морось"
        case 61, 63, 65, 66, 67, 80, 81, 82: return "Дождь"
        case 71, 73, 75, 77, 85, 86: return "Снег"
        case 95, 96, 99: return "Гроза"
        default: return "—"
        }
    }

    private func requestData(from url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.timeoutInterval = 15
        return try await session.data(for: request)
    }
}

func iosWeatherTripDateRange(_ value: String) -> (Date, Date)? {
    let normalizedValue = value
        .replacingOccurrences(of: "\u{00A0}", with: " ")
        .replacingOccurrences(of: "\u{202F}", with: " ")
        .replacingOccurrences(of: "\u{2009}", with: " ")
        .trimmingCharacters(in: .whitespacesAndNewlines)
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .current

    let isoFormatter = DateFormatter()
    isoFormatter.locale = Locale(identifier: "en_US_POSIX")
    isoFormatter.calendar = calendar
    isoFormatter.timeZone = calendar.timeZone
    isoFormatter.dateFormat = "yyyy-MM-dd"
    let isoDates = weatherRegexMatches(#"(?<!\d)\d{4}-\d{2}-\d{2}(?!\d)"#, in: normalizedValue)
        .compactMap { isoFormatter.date(from: $0) }
    if let first = isoDates.first {
        return (first, isoDates.dropFirst().first ?? first)
    }

    let dottedFormatter = DateFormatter()
    dottedFormatter.locale = Locale(identifier: "en_US_POSIX")
    dottedFormatter.calendar = calendar
    dottedFormatter.timeZone = calendar.timeZone
    dottedFormatter.dateFormat = "dd.MM.yyyy"
    let dottedDates = weatherRegexMatches(#"(?<!\d)\d{1,2}[./]\d{1,2}[./]\d{4}(?!\d)"#, in: normalizedValue)
        .compactMap { dottedFormatter.date(from: $0.replacingOccurrences(of: "/", with: ".")) }
    if let first = dottedDates.first {
        return (first, dottedDates.dropFirst().first ?? first)
    }

    let humanRangePattern = #"(?<!\d)(\d{1,2})\s*[–—-]\s*(\d{1,2})\s+([A-Za-zА-Яа-яЁёÄÖÜäöüß]+)\s+(\d{4})(?!\d)"#
    if let match = weatherRegexCaptureMatches(humanRangePattern, in: normalizedValue).first,
       match.count == 5,
       let startDay = Int(match[1]),
       let endDay = Int(match[2]),
       let year = Int(match[4]),
       let month = iosWeatherMonthNumber(match[3]),
       let start = calendar.date(from: DateComponents(year: year, month: month, day: startDay)),
       let end = calendar.date(from: DateComponents(year: year, month: month, day: endDay)) {
        return (start, end)
    }

    let humanPattern = #"(?<!\d)(\d{1,2})\s+([A-Za-zА-Яа-яЁёÄÖÜäöüß]+)\s+(\d{4})(?!\d)"#
    let humanDates = weatherRegexCaptureMatches(humanPattern, in: normalizedValue).compactMap { match -> Date? in
        guard match.count == 4,
              let day = Int(match[1]),
              let year = Int(match[3]),
              let month = iosWeatherMonthNumber(match[2])
        else { return nil }
        return calendar.date(from: DateComponents(year: year, month: month, day: day))
    }
    guard let first = humanDates.first else { return nil }
    return (first, humanDates.dropFirst().first ?? first)
}

func iosWeatherTripDates(_ value: String, maxDays: Int = 366) -> [Date] {
    guard let (start, end) = iosWeatherTripDateRange(value) else { return [] }
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .current
    let dayCount = calendar.dateComponents([.day], from: start, to: end).day ?? 0
    guard dayCount >= 0, maxDays > 0 else { return [] }
    let lastOffset = min(dayCount, maxDays - 1)
    return (0...lastOffset).compactMap {
        calendar.date(byAdding: .day, value: $0, to: start)
    }
}

func iosWeatherISODate(_ date: Date) -> String {
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter.string(from: date)
}

private func iosWeatherMonthNumber(_ value: String) -> Int? {
    switch value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased().prefix(3) {
    case "янв", "jan", "ene": return 1
    case "фев", "feb": return 2
    case "мар", "mär": return 3
    case "апр", "apr": return 4
    case "мая", "май", "may", "mai": return 5
    case "июн", "jun": return 6
    case "июл", "jul": return 7
    case "авг", "aug", "ago": return 8
    case "сен", "сент", "sep": return 9
    case "окт", "oct", "okt": return 10
    case "ноя", "nov": return 11
    case "дек", "dec", "dez": return 12
    default: return nil
    }
}

private func weatherRegexMatches(_ pattern: String, in value: String) -> [String] {
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
    let range = NSRange(value.startIndex..., in: value)
    return regex.matches(in: value, range: range).compactMap { match in
        guard let matchRange = Range(match.range, in: value) else { return nil }
        return String(value[matchRange])
    }
}

private func weatherRegexCaptureMatches(_ pattern: String, in value: String) -> [[String]] {
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
    let range = NSRange(value.startIndex..., in: value)
    return regex.matches(in: value, range: range).map { match in
        (0..<match.numberOfRanges).compactMap { index in
            guard let matchRange = Range(match.range(at: index), in: value) else { return nil }
            return String(value[matchRange])
        }
    }
}

private enum WeatherError: Error {
    case requestFailed
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

private extension Array where Element: Hashable {
    func uniqued() -> [Element] {
        var seen = Set<Element>()
        return filter { seen.insert($0).inserted }
    }
}
